### PR TITLE
fix: fallback to super.getLocale() if no custom locale is set

### LINKED ISSUE #2379: getLocale on DatePicker overwritten
### Description

The default implementation of `getLocale` in `com.vaadin.flow.component.Component` does return the locale of the current UI so developers are used to be able of just using component.getLocale() to get the locale from a component. It's also used for other internal elements like translation etc.
`com.vaadin.flow.component.datepicker.DatePicker` overwrites `getLocale` with `return locale;` which initially is null and gets initiated with the locale of the UI on the onAttach event. This will mean, that getLocale() of a datePicker returns null for a while getLocale of every other component returns the currentUI locale. I would consider this as a bug in the logic and even if it is not, it might lead to potential NPE as it already did in our application. 
I recommand to overwrite getLocale in DatePicker in a way which returns the locale variable if it is not null and returns super.getLocale otherwise.
`DateTimePicker` does it a little bit differently here and fills `locale` with the currentUIs locale in the constructor which is different again to the DatePicker implementation. I would here - as well - remove the setting of locale in the constructor and overwrite the getLocale with the use of if (locale == null) return super.getLocale()

- Seems like the new getLocale implementation happened here: https://github.com/vaadin/vaadin-date-picker-flow/pull/54
- And this is the commit where the setLocale from the constructor was removed, leading to NPEs in our code: https://github.com/vaadin/flow-components/commit/863e2b0a1fa376f4d309468096403f279e504ae3 

### Expected outcome

getLocale() of DatePicker should return the currentUI locale as every other component if locale is not set

### Actual outcome

getLocale() of DatePicker returns null

### Minimal reproducible example

```
class PublicLocaleTextField extends TextField {
	@Override
	public Locale getLocale() {
		return super.getLocale();
	}
}

PublicLocaleTextField textField = new PublicLocaleTextField();
Objects.requireNonNull(textField.getLocale()); // no exception

DatePicker datePicker = new DatePicker();
Objects.requireNonNull(datePicker.getLocale());  // throws NullPointerException
```

### Steps to reproduce

1. Create a DatePicker
2. Call getLocale() before onAttach function is called
3. Create your own component which makes getLocale public or just uses it inside of the component
4. Call getLocale() on or in this component before onAttach function is called

### Environment

- Vaadin 14.7.5

### Browsers Affected
- [x ] Chrome
- [x] Firefox
- [x] Safari
- [x] Edge
- [x] IE 11
- [x] iOS Safari
- [x] Android Chrome

### PR DESCRIPTION
**Pull request by sissbruecker**
## Description

Extended version of #2380 that covers DatePicker, TimePicker and DateTimePicker. Also refactors the components to pull the locale on demand, rather than setting the custom locale fields to the UI locale on attach.

Opening this as a separate PR for now, in case we don't want to go ahead with the refactoring.

Fixes #2379 

## Type of change

- [x] Bugfix

### CODE REVIEW

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 1]
```diff
@@ -329,21 +314,18 @@ public void setLocale(Locale locale) {
...
         initConnector();
-        if (locale == null) {
-            getUI().ifPresent(ui -> setLocale(ui.getLocale()));
-        } else if (languageTag != null) {
-            requestI18nUpdate();
-        }
-        if (i18n != null) {
-            requestI18nUpdate();
-        }
+        requestI18nUpdate();
```
Comment (sissbruecker) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:328]: All of the cases above should trigger `requestI18nUpdate()`, setting the custom locale field to UI locale is not necessary anymore.

File: vaadin-date-time-picker-flow-parent/vaadin-date-time-picker-flow/src/main/java/com/vaadin/flow/component/datetimepicker/DateTimePicker.java [hunk 2]
```diff
@@ -148,8 +148,6 @@ public DateTimePicker(LocalDateTime initialDateTime) {
         addToSlot(datePicker, "date-picker");
         addToSlot(timePicker, "time-picker");
 
-        setLocale(UI.getCurrent().getLocale());
```
Comment (sissbruecker) [inline vaadin-date-time-picker-flow-parent/vaadin-date-time-picker-flow/src/main/java/com/vaadin/flow/component/datetimepicker/DateTimePicker.java:151]: Setting the custom locale field to UI locale is not necessary anymore, DatePicker and TimePicker will automatically use the current UI locale internally.

File: vaadin-time-picker-flow-parent/vaadin-time-picker-flow-integration-tests/src/test/java/com/vaadin/flow/component/timepicker/tests/TimePickerDetachAttachPageIT.java [hunk 3]
```diff
@@ -15,36 +15,61 @@
...
         // Detaching and attaching time picker
-        WebElement toggleAttach = findElement(By.id("toggle-attached"));
         toggleAttach.click();
         toggleAttach.click();
 
         assertTimePickerIsValidOnTab();
     }
 
+    @Test
+    public void formatShouldRespectLocaleAfterDetachAndReattach() {
```
Comment (sissbruecker) [inline vaadin-time-picker-flow-parent/vaadin-time-picker-flow-integration-tests/src/test/java/com/vaadin/flow/component/timepicker/tests/TimePickerDetachAttachPageIT.java:57]: This PR also fixes a bug in the TimePicker where the time format from the locale would not be reapplied when detaching and reattaching. See changes in `TimePicker.onAttach`.

File: vaadin-time-picker-flow-parent/vaadin-time-picker-flow/src/main/java/com/vaadin/flow/component/timepicker/TimePicker.java [hunk 4]
```diff
@@ -59,11 +62,11 @@
     private static final String PROP_AUTO_OPEN_DISABLED = "autoOpenDisabled";
 
     private Locale locale;
-    private transient DateTimeFormatter dateTimeFormatter;
```
Comment (sissbruecker) [inline vaadin-time-picker-flow-parent/vaadin-time-picker-flow/src/main/java/com/vaadin/flow/component/timepicker/TimePicker.java:62]: Unused code, same with `initializeAndReturnFormatter` below.

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 5]
```diff
@@ -408,6 +390,24 @@ private void executeI18nUpdate() {
...
         }
 
+        // For ill-formed locales, Locale.toLanguageTag() will append subtag
+        // "lvariant" to it, which will cause the client side
+        // Date().toLocaleDateString()
+        // fallback to the system default locale silently.
+        // This has been caught by DatePickerValidationPage::invalidLocale test
+        // when running on
+        // Chrome(73+)/FireFox(66)/Edge(42.17134).
+        Locale appliedLocale = getLocale();
```
Comment (sissbruecker) [inline vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java:400]: Locale is now pulled on demand, when updating I18N settings in connector.

File: vaadin-time-picker-flow-parent/vaadin-time-picker-flow/src/main/java/com/vaadin/flow/component/timepicker/TimePicker.java [hunk 6]
```diff
@@ -416,7 +407,37 @@ public void setLocale(Locale locale) {
...
+            }
+            pendingLocaleUpdate = ui.beforeClientResponse(this, context -> {
+                pendingLocaleUpdate = null;
+                executeLocaleUpdate();
+            });
+        });
+    }
+
+    private void executeLocaleUpdate() {
+        Locale appliedLocale = getLocale();
```
Comment (sissbruecker) [inline vaadin-time-picker-flow-parent/vaadin-time-picker-flow/src/main/java/com/vaadin/flow/component/timepicker/TimePicker.java:430]: Locale is now pulled on demand, when updating locale settings in connector.

**Review by hdamr — APPROVED** [review]
Looks good.

**Pull request merged**